flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java

```diff
@@ -137,7 +137,13 @@ public static long getSizeOfFreeHeapMemoryWithDefrag() {
 	 */
 	public static long getSizeOfFreeHeapMemory() {
 		Runtime r = Runtime.getRuntime();
-		return r.maxMemory() - r.totalMemory() + r.freeMemory();
+		long maxMemory = r.maxMemory();
+		if (maxMemory == Long.MAX_VALUE) {
+			// workaround for some JVM versions
+			return r.freeMemory();
```

Comment:
I haven't tested this yet, but I think this [will not work](http://stackoverflow.com/questions/12807797/java-get-available-memory), because `freeMemory` returns to the free memory from the currently allocated memory. The JVM allocates memory in chunks. Free memory refers to how much of these allocated chunks is free.